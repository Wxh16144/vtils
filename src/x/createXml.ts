import { j2xParser as J2XParser, J2xOptionsOptional } from 'fast-xml-parser'

const attrTagName = '@__attr__@' as const
const textTagName = '@__text__@' as const
const cdataTagName = '@__cdata__@' as const

/**
 * 创建 XML 文本。
 *
 * @param data 数据
 * @param options 选项
 */
export function createXml(data: any, options?: J2xOptionsOptional): string {
  return new J2XParser({
    ...options,
    attrNodeName: attrTagName,
    textNodeName: textTagName,
    cdataTagName: cdataTagName,
  }).parse(data)
}

createXml.attr = function <T = any>(value: T): any {
  return { [attrTagName]: value }
}

createXml.text = function <T = any>(value: T): any {
  return { [textTagName]: value }
}

createXml.cdata = function <T = any>(value: T): any {
  return { [cdataTagName]: value }
}
