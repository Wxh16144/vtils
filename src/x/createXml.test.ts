import { createXml } from './createXml'

describe('createXml', () => {
  test('表现正常', () => {
    expect(
      createXml({
        id: 1,
        name: 'jay',
        close: false,
        content: createXml.cdata('大家好 > kkkdkkdkd ?'),
        desc: {
          ...createXml.attr({
            time: '2020',
          }),
          ...createXml.text('我是xxxx'),
        },
      }),
    ).toMatchSnapshot()
  })
})
