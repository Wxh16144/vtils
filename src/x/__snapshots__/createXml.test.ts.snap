// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createXml 表现正常 1`] = `
<id>
  1
</id>
<name>
  jay
</name>
<close>
  false
</close>
<content>
  <!--[CDATA[大家好 > kkkdkkdkd ?]]-->
</content>
<desc time="2020">
  我是xxxx
</desc>
<list>
  <item>
    <id>
      1
    </id>
  </item>
  <item>
    <id>
      2
    </id>
  </item>
</list>
`;
